File: leetcode/solved/7_Reverse_Integer/solution.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2018 sungmin <smoh2044@gmail.com>
#
# Distributed under terms of the MIT license.

"""
Given a 32-bit signed integer, reverse digits of an integer.

Example 1:
Input: 123
Output:  321

Example 2:
Input: -123
Output: -321

Example 3:
Input: 120
Output: 21

Note:
Assume we are dealing with an environment which could only hold integers within the 32-bit signed
integer range. For the purpose of this problem, assume that your function returns 0 when the
reversed integer overflows.
"""


class Solution:
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x == 0:
            return 0
        neg = 1
        if x < 0:
            neg = -1
            x *= -1
        lst = list(str(x))
        lst.reverse()
        x = int(''.join(lst))
        if x > 2147483648:
            return 0
        return neg * x


def main():
    print(Solution().reverse(int(input())))


if __name__ == '__main__':
    main()

File: leetcode/solved/7_Reverse_Integer/solution.cpp
/*
 * solution.cpp
 * Copyright (C) 2018 sungmin <smoh2044@gmail.com>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>


using namespace std;


int reverse(int x) {
    if(x == 0){
        return 0;
    }
    int neg = 1;
    if(x < 0){
        neg = -1;
        x *= -1;
    }
    long i = 1;
    for(; i <= x; i *= 10){}
    i /= 10;
    long ret = 0;
    for(; i > 0; i /= 10, x /= 10){
        ret += (i * (x%10));
        if(ret > INT_MAX){
            return 0;
        }
    }
    return neg * ret;
}


int main(){
    int x;
    scanf("%d", &x);
    int y = reverse(x);
    printf("%d\n", y);
    return 0;
}
